src/lib.rs

```diff
@@ -37,17 +37,16 @@
 //! are already somewhat familiar with Rust, and who want to see something larger and more involved
 //! be built. You can find the recordings of past sessions [on
 //! YouTube](https://www.youtube.com/playlist?list=PLqbS7AVVErFgY2faCIYjJZv_RluGkTlKt).
+#![warn(unreachable_pub)]
+#![warn(missing_copy_implementations)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unused_extern_crates)]
```

Comment:
Also
```rust
#![warn(rust_2018_idioms)]
#![warn(missing_debug_implementations)]
```
We should also probably run `clippy` (if you're not already).

Reply:
There are 4 clippy complaints which I have chosen not to address, all of which are "very complex type used", e.g., 

```
warning: very complex type used. Consider factoring parts into `type` definitions
  --> src/providers/azure.rs:54:9
   |
54 |         Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity
```

Reply:
Just add
```
#![allow(clippy::type_complexity)]
```
(I think that's the right invocation)